### PR TITLE
Don't release ticket reservations for transactions that are in…

### PR DESCRIPTION
**Pull request by mnelson4**
## Problem this Pull Request solves

If the cron task `AHEE__EE_Cron_Tasks__clean_up_junk_transactions` ran while a transaction was in progress and had tickets reserved, those reservations were getting released, would could lead to overselling. See https://github.com/eventespresso/event-espresso-core/issues/2047#issuecomment-566655969

I vaguely remember that over the past few years we've had some overselling issues. The problem this is fixing may have been the cause of them. (Specifically, that there were reservations on the ticket for registrations in progress, then the cron task would run and release the ticket reservations, and before the registrations finished up, someone else would start to register.)

As far as internal details of what changed, it's a bit confusing...
The logic in `\EED_Ticket_Sales_Monitor::reset_reservation_counts` seems to have gotten confused at some point.
There were two problems:

### 1. PHP array union overwrites based on array keys
The list of `$valid_reserved_tickets` was being built using `$valid_reserved_tickets += EED_Ticket_Sales_Monitor::get_ticket_line_items_for_grand_total($total_line_item            );`. That operator `+=` union'd `$valid_reserved_tickets` with `EED_Ticket_Sales_Monitor::get_ticket_line_items_for_grand_total()`'s results based on array **keys**. And the array was indexed **numerically**. So if `$valid_reserved_tickets` had one item, its array key was `0`. If `EED_Ticket_Sales_Monitor::get_ticket_line_items_for_grand_total()` returned an array which also had a single item, its key was also `0`. So when they were union'd, the second array's `0` was ignored.
> The + operator returns the right-hand array appended to the left-hand array; for keys that exist in both arrays, the elements from the left-hand array will be used, and the matching elements from the right-hand array will be ignored. 
from https://www.php.net/manual/en/language.operators.array.php

So, even if there were 400 `$transactions_not_in_progress`, the `$valid_reserved_tickets` derived from them was usually just a single item.

The fix was to instead index the array using the line item's ID.

### 2. Fetching Expired Transactions instead of In-Progress Ones
`\EED_Ticket_Sales_Monitor::reset_reservation_counts` was fetching the list of expired transactions (which could never complete, and so their reservations on tickets could be released) but then it considered their tickets' reservations as "valid" by adding them to `$valid_reserved_tickets` 😵. That seems to be the opposite of what we want: the transactions are **expired**, and so their ticket reservations are **no longer** valid. 

The fix was to do the opposite: fetch the list of transactions that were still **in-progress** (ie, their session hadn't yet expired) and could be completed. Those are the transactions whose ticket reservations are still valid because they could still complete their registrations.

So in summary, on this branch, we are fetching all the in-progress transactions, and building the list of `$valid_reserved_tickets` from that, indexing the items by the ticket line item's ID in the array (rather than numerically) in order for the array `+=` operator to work as expected.

### Regarding efficiency
I originally dug into this code to see if there was a way to optimize it further. I got sidetracked fixing this bug, but I think this may also help significantly with efficiency too.
Before we were fetching the list all expired transactions (which hadn't yet been deleted). So that could have been a week's worth of transactions being fetched on every execution of the cron job. On some sites, there were hundreds or even thousands of these.
With these changes, we're instead only grabbing all the in-progress transactions. While on some sites this number can also be high, it's at least only an hour's worth of transactions instead of a week's.
So I'm guessing this means that where before we were fetching hundreds or thousands of transactions everytime the cron ran, we'll now instead be fetching dozens, so the load will be alleviated.

## How has this been tested

Besides the new unit test, 
* [x] Create an event, set a limit to 10 on its ticket
* Register 10 spots for the event in Firefox, but don't actually answer any reg details. Keep the tab open though. Notice the ticket's "Rsvd" count is now at 10.
* Execute the cron task AHEE__EE_Cron_Tasks__clean_up_junk_transactions (I use the WP CLI command wp cron event run AHEE__EE_Cron_Tasks__clean_up_junk_transactions)
* Notice that the event's "Rsvd" count is back to 0 (even though there is an active transaction for 10 of them!!!)
* In Google Chrome, go to the event and register another 10 spots. Complete the registration.
* Complete the registration you started in Firefox.
* Notice the event now has 20 tickets sold, despite having a limit of 10 on the ticket skull_and_crossbones skull_and_crossbones skull_and_crossbones. Here's a video of that on master: drive.google.com/file/d/1emSMmhy1sSPhvFaL7w6cvDT4U4MqaJnM/view
* [ ] Create another event (or two) with various tickets. Make a few registrations and have a few registrations in progress, so some of the tickets have tickets reserved. 
* Go to the maintenance -> "Reset/Delete Data" and click "Reset Ticket and Datetime Reserved Counts". The tickets' reserved counts should NOT change (on master, they'd all get set back to 0.)
* [x] Create a ticket with a limit of 10. Begin 10 registrations for that ticket, but don't answer any reg questions. So, the ticket's fully reserved.
* In another browser, try to register for that ticket. You should be prevented from doing so.
* In the DB, manually change that first transaction's `TXN_timestamp` to anything older than a week, and then run the cron task again.
* The ticket's reserved count should be set back to 0

## Checklist

* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [x] User input is adequately validated and sanitized
* [x] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [x] My code is tested.
* [x] My code follows the Event Espresso code style.
* [x] My code has proper inline documentation.
* [x] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

**Comment by mnelson4** [comment]
Update: In Slack, Brent and I discussed a little. Brent said
> i think i probably should not have named that var `$valid_reserved_tickets` because that gets subtracted from the total of reserved tickets
> ie: those get released (edited) 
> look at `\EED_Ticket_Sales_Monitor::release_reservations_for_tickets()`
> you've got this:
> `$reserved_qty = $ticket_with_reservations->reserved();`
> and then this:
> `$reserved_qty -= $valid_reserved_ticket_line_item->quantity();`
> ya it was bad naming on my behalf

Which led to me think the above changes weren't correct...

BUT, when I investigated more, I saw

```
$ticket_with_reservations->decreaseReserved($reserved_qty,...);
```

SO, `$reserved_qty` isn't the number of valid reserved spots we're setting `TKT_reserved` TO, it's the number of reserved spots WE'RE RELEASING. The opposite.
So there's a double-negative here: `$valid_reserved_ticket_line_item->quantity()` is reducing the number to release.
So, "`$valid_reserved_tickets` gets subtracted from the total of reserved tickets" isn't correct; if there are no `$valid_reserved_tickets`, then all that ticket's reservations get released. But if there are `$valid_reserved_tickets`, we reduce the number to release (or if there are enough valid reserved tickets, we don't release any).

SO, I'm really sorry, I know this is as frustrating as it is confusing, these changes again appear to be correct. I added a test to verify `\EED_Ticket_Sales_Monitor::release_reservations_for_tickets` worked as I expected (namely, that `$valid_reserved_ticket_line_items`'s name is actually true to its purpose, as all other tickets reservations get released unless they have valid reserved ticket line items), and some comments to try to justify why the code works.

Lastly, I noticed an unnecessary save happening in `\EED_Ticket_Sales_Monitor::release_reservations_for_tickets` (we just want to update the ticket's `TKT_reserved` count, which `\EE_Ticket::decreaseReserved` already does) so I removed it.

File: modules/ticket_sales_monitor/EED_Ticket_Sales_Monitor.module.php [hunk 1]
```diff
@@ -960,70 +964,72 @@ private static function get_ticket_line_items_for_grand_total(EE_Line_Item $tota
...
-            if (self::debug) {
-                echo self::$nl . ' . $ticket_with_reservations->ID(): ' . $ticket_with_reservations->ID();
-                echo self::$nl . ' . $reserved_qty: ' . $reserved_qty;
-            }
+            // The $valid_reserved_ticket_line_items tells us what the reserved count on their tickets (and datetimes)
+            // SHOULD be. Instead of just directly updating the list, we're going to use EE_Ticket::decreaseReserved()
+            // to try to avoid race conditions, so instead of just finding the number to update TO, we're going to find
+            // the number to RELEASE. It's the same end result, just different path.
+            // Begin by assuming we're going to release all the reservations on this ticket.
+            $num_tix_for_expired_reservations = $ticket_with_reservations->reserved();
```
Comment (tn3rb) [inline modules/ticket_sales_monitor/EED_Ticket_Sales_Monitor.module.php]: how about just `$expired_reservations` or `$expired_reservations_count` or `$expired_reserved_qty`???

Comment (mnelson4) [reply modules/ticket_sales_monitor/EED_Ticket_Sales_Monitor.module.php]: Yeah good suggestions IMO

**Comment by joshfeck** [comment]
I tested this and it looks good to go. 

I also tested what would happen after one hour when the transaction's AHEE__EE_Cron_Tasks__expired_transaction_check was fired and found that the reserved counts were cleared, but that I could still complete the transaction as it seemed to still have an open session. Which could lead to overselling if someone else happened to register in between the time the AHEE__EE_Cron_Tasks__expired_transaction_check was run and the other transaction was completed. This is also an issue on master, so not a regression due to the work on this branch.

**Comment by mnelson4** [comment]
> I tested this and it looks good to go.

Thanks, merged!

> I also tested what would happen after one hour when the transaction's AHEE__EE_Cron_Tasks__expired_transaction_check was fired and found that the reserved counts were cleared, but that I could still complete the transaction as it seemed to still have an open session

Hmmmm ya I'll make an issue for that...

**Pull request merged**